src/esp/CMakeLists.txt

```diff
@@ -17,15 +17,595 @@ if(BUILD_WARNINGS_AS_ERRORS)
+  sim/SimulatorConfiguration.h
+)
+
+#################
+## build library
+
+################################
+# Keeping gfx_batch a separate library
 add_subdirectory(gfx_batch)
-add_subdirectory(assets)
-add_subdirectory(metadata)
-add_subdirectory(io)
-add_subdirectory(scene)
-add_subdirectory(physics)
-add_subdirectory(nav)
-add_subdirectory(sensor)
-add_subdirectory(sim)
+##############################
+
+set(habsim_SOURCES ${core_SOURCES})
+list(
+  APPEND
+  habsim_SOURCES
+  ${geo_SOURCES}
+  ${gfx_SOURCES}
+  ${assets_SOURCES}
+  ${metadata_SOURCES}
+  ${io_SOURCES}
+  ${scene_SOURCES}
+  ${physics_SOURCES}
+  ${nav_SOURCES}
+  ${sensor_SOURCES}
+  ${sim_SOURCES}
+)
+
+add_library(
+  habsim STATIC
+  ${habsim_SOURCES}
+)
+
+target_include_directories(
+  habsim PUBLIC ${PROJECT_BINARY_DIR}
+  PRIVATE "${DEPS_DIR}/tinyxml2" "${DEPS_DIR}/recastnavigation/Detour/Include"
+          "${DEPS_DIR}/recastnavigation/Recast/Include"
+)
+
+if(BUILD_WITH_CUDA)
+  add_library(noise_model_kernels STATIC ${noise_model_SOURCES})
+  target_link_libraries(noise_model_kernels PUBLIC ${CUDART_LIBRARY})
+  target_include_directories(
+    noise_model_kernels PRIVATE ${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}
+  )
+  target_compile_features(noise_model_kernels PUBLIC cxx_std_11)
+
+  target_link_libraries(habsim PUBLIC noise_model_kernels ${CUDART_LIBRARY})
+
+  target_include_directories(habsim PRIVATE ${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES})
+endif()
+
+if(BUILD_WITH_AUDIO)
+  include_directories(${DEPS_DIR}/rlr-audio-propagation/RLRAudioPropagationPkg/headers)
+  target_link_libraries(habsim PUBLIC ${RLRAudioPropagation_LIBRARY})
+endif()
+
+target_link_libraries(
+  habsim
+  PUBLIC gfx_batch
+         Magnum::AnyImageImporter
+         Magnum::AnySceneImporter
+         Magnum::GL
+         Magnum::Magnum
+         Magnum::MaterialTools
+         Magnum::MeshTools
+         Magnum::DebugTools
+         Magnum::SceneGraph
+         Magnum::SceneTools
+         Magnum::Shaders
+         Magnum::Trade
+         Magnum::Primitives
+         MagnumPlugins::BasisImporter
+         MagnumPlugins::GltfImporter
+         MagnumPlugins::PrimitiveImporter
+         MagnumPlugins::StanfordImporter
+         MagnumPlugins::StbImageImporter
+         MagnumPlugins::StbImageConverter
+         MagnumIntegration::Eigen
+         Corrade::Utility
+         Magnum::AnyImageConverter
+  PRIVATE tinyxml2 Detour Recast
+)
+
+# Link windowed application library if needed
+if(BUILD_GUI_VIEWERS)
+  target_link_libraries(habsim PUBLIC Magnum::GlfwApplication)
+endif()
+
+if(BUILD_WITH_BACKGROUND_RENDERER)
+  target_link_libraries(habsim PUBLIC atomic_wait)
+endif()
+
+if(BUILD_ASSIMP_SUPPORT)
+  target_link_libraries(
+    habsim PUBLIC MagnumPlugins::AssimpImporter PRIVATE Assimp::Assimp
+  )
+endif()
+
+if(OpenMP_CXX_FOUND)
+  target_link_libraries(habsim PUBLIC OpenMP::OpenMP_CXX)
+endif()
+
+target_link_libraries(
+  habsim
+  PUBLIC Magnum::WindowlessApplication
+)
```

Comment:
On the other hand, this one is fine, the windowless application _is_ used inside the library.